client/packages/system/src/Immunisation/VaccineCourseEditModal/VaccineCourseEditModal.tsx

```diff
@@ -0,0 +1,324 @@
+    draft,
+    update: { update },
+    create: { create },
+    updatePatch,
+    query: { isLoading },
+    isDirty,
+    setIsDirty,
+  } = useVaccineCourse(vaccineCourse?.id ?? undefined);
+  const { data: demographicData } = useDemographicIndicators();
+
+  const { Modal } = useDialog({ isOpen, onClose, disableBackdrop: true });
+  const height = useKeyboardHeightAdjustment(600);
+
+  // const defaultRow: VaccineCourseScheduleNode = {
+  //   doseNumber: 1,
+  //   id: FnUtils.generateUUID(),
+  //   label: '',
+  //   __typename: 'VaccineCourseScheduleNode',
+  // };
+  const tryUpdateValue = (value: number | undefined) => {
+    if (typeof value !== 'number') {
+      return;
+    }
+    updatePatch({ doses: value });
+  };
+
+  // const updateSchedule = (value: number) => {
+  //   if (!value) {
+  //     return;
+  //   }
+  //   const scheduleSeed = (number: number): VaccineCourseScheduleNode => {
+  //     return {
+  //       __typename: 'VaccineCourseScheduleNode',
+  //       id: FnUtils.generateUUID(),
+  //       doseNumber: number,
+  //       label: '',
+  //     };
+  //   };
+  //   let rows = draft?.vaccineCourseSchedules ?? [];
+
+  //   if (rows.length === value) {
+  //     return;
+  //   } else if (value > rows.length) {
+  //     let toAdd = value - rows.length;
+  //     while (toAdd > 0) {
+  //       const number = value - toAdd + 1;
+  //       rows.push(scheduleSeed(number));
+  //       toAdd--;
+  //     }
+  //   } else {
+  //     rows = rows.slice(0, value);
+  //   }
+  //   updatePatch({ vaccineCourseSchedules: rows });
+  // };
+
+  // const updateDescription = (patch: RecordPatch<VaccineCourseScheduleNode>) => {
+  //   if (!patch) {
+  //     return;
+  //   }
+  //   const schedule = {
+  //     ...ArrayUtils.toObject(draft.vaccineCourseSchedules ?? [defaultRow]),
+  //     [patch.id]: patch,
+  //   };
+
+  //   const scheduleArray = Object.values(
+  //     schedule
+  //   ) as VaccineCourseScheduleNode[];
+  //   updatePatch({ vaccineCourseSchedules: scheduleArray });
+  // };
+
+  // const dosesColumns = useColumns<VaccineCourseScheduleNode>(
+  //   [
+  //     { key: 'doseNumber', label: 'label.dose-number' },
+  //     [descriptionColumn(t('label.new-row')), { setter: updateDescription }],
+  //   ],
+  //   {},
+  //   [draft]
+  // );
+
+  const options = useMemo(
+    () => getDemographicOptions(demographicData?.nodes ?? []),
+    [demographicData]
+  );
+
+  if (isLoading) {
+    return <BasicSpinner />;
+  }
+
+  const defaultValue = {
+    value: draft.demographicIndicator?.name ?? '',
+    label: draft.demographicIndicator
+      ? `${draft.demographicIndicator?.name} ${draft.demographicIndicator?.baseYear}`
+      : '',
+  };
+
+  return (
+    <Modal
+      title=""
+      cancelButton={<DialogButton variant="cancel" onClick={onClose} />}
+      okButton={
+        <DialogButton
+          disabled={!isDirty || !programId}
+          variant="ok"
+          onClick={async () => {
```

Comment:
I'd lift this out to a `save` function - it's hard to read it all inline like this.

The error handling could also have an extra case - currently if the typename isn't `VaccineCourseNode` then the modal doesn't close, and no message is shown